0027-safely-recommend-actions.adoc

```diff
@@ -0,0 +1,173 @@
+
+2. A pair of expiries is useful if there is time to take each action required to complete the swap i.e., the expiry pair is useful if there is time to transition to the first action AND there is time for each subsequent action.
+
+3. Semantically the next possible action is either the next happy path action or cancel. Cancel is either refund or abort.
+
+.Semantic actions
+* *next*: The next happy path action (deterministic from the state diagram).
+* *refund*: Send a refund transaction to the blockchain (cancel after fund).
+* *wait_for_refund*: Refund is not yet available but should be taken when available.
+* *abort*: Take no further action (cancel before fund).
+
+The recommended next action can be fully determined from the following algorithm:
+
+----
+1. IF refund is available RETURN refund
+2. IF there is time to transition to next AND there is time for the subsequent action RETURN next
+3. IF have funded RETURN wait_for_refund
+4. RETURN abort
+----
+
+Note that step 2 is recursive, the base case being the final stage of the annotated state diagram - done (see below).
+
+=== Explanation
+
+Various events must occur in order to transition from one state to the next. We can identify these events and annotate the state diagram accordingly. If we are able to calculate the time of each of these events we are able to fully determine the period of each transition.
+
+We define the Time To Final to be a part of the transition to a state e.g., the alpha funded state is not reached until the alpha fund transaction has become final.
+
+There exists a transition diagram for each actor. Each actor progresses along their own state diagram but must be able to predict the state progression of the counterparty in order to determine whether progression is possible for the counterparty.
+
+.Annotated state diagram for Alice:
+----
+(Create) TTA_A TTI_AF TTF_A (alpha funded) TTA_B TTI_BF TTF_B (beta funded) TTA_A TTI_BR TTC_B (beta redeemed) (done)
+----
+
+.Annotated state diagram for Bob:
+----
+(Create) TTA_A TTI_AF TTF_A (alpha funded) TTA_B TTI_BF TTF_B (beta funded) TTA_A TTI_BR (beta redeemed) TTA_B TTI_AR TTC_A (alpha redeemed) (done)
+----
+
+For a swap to continue, at all times, each actor must be able to transition to 'done' before the expiry time governing their redeem action elapses.
+
+.Key:
+* (state): states are shown inside parenthesis
+* `TTA_A`: `TTA` for Alice
+* `TTA_B`: `TTA` for Bob
+* `TTI_AF`: `TTI` for Alpha Fund transaction
+* `TTI_BF`: `TTI` for Beta Fund transaction
+* `TTI_AR`: `TTI` for Alpha Redeem transaction
+* `TTI_BR`: `TTI` for Beta Redeem transaction
+* `TTF_A`: `TTF` for Alpha ledger
+* `TTF_B`: `TTF` for Beta ledger
+
+.Note
+* `TTI_AR` is not actually essential for Bob, we assume Bob waits for Alice's redeem transaction to be included in a block, technically by monitoring unconfirmed transactions (e.g. watching the Bitcoin mempool) `TTI_AR` could be removed. This in no way effects the security of the protocol.
```

Comment:
```suggestion
* `TTI_AR` is not actually essential for Bob, we assume Bob waits for Alice's redeem transaction to be included in a block, technically by monitoring unconfirmed transactions (e.g. watching the Bitcoin mempool) `TTI_AR` could be removed. This in no way effects the security of the protocol.
```
Adding a note on why this is not essential could help the reader.

Reply:
I thought this was the note :)